overlord/devicemgmtstate/devicemgmtmgr.go

```diff
@@ -0,0 +1,561 @@
+
+	return m, nil
+}
+
+// getState retrieves the current device management state, initializing if not present.
+func (m *DeviceMgmtManager) getState() (*DeviceMgmtState, error) {
+	var ms DeviceMgmtState
+	err := m.state.Get(deviceMgmtStateKey, &ms)
+	if err != nil {
+		if errors.Is(err, state.ErrNoState) {
+			return &DeviceMgmtState{
+				PendingMessages: make(map[string]*PendingMessage),
+				Sequences:       make(map[string]*Sequence),
+				ReadyResponses:  make(map[string]store.Message),
+			}, nil
+		}
+
+		return nil, err
+	}
+
+	return &ms, nil
+}
+
+// setState persists the device management state.
+func (m *DeviceMgmtManager) setState(ms *DeviceMgmtState) {
+	m.state.Set(deviceMgmtStateKey, ms)
+}
+
+// Ensure implements StateManager.Ensure.
+func (m *DeviceMgmtManager) Ensure() error { // TODO: add manager-level feature flag?
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	ms, err := m.getState()
+	if err != nil {
+		return err
+	}
+
+	for _, chg := range m.state.Changes() {
+		if chg.Kind() == deviceMgmtCycleChangeKind && !chg.Status().Ready() {
+			return nil
+		}
+	}
+
+	ts := state.NewTaskSet()
+	var dispatch *state.Task
+	for _, msg := range ms.PendingMessages {
+		if msg.ChangeID == "" {
+			dispatch = m.state.NewTask("dispatch-messages", "Dispatch message(s) to subsystems")
+			ts.AddTask(dispatch)
+			break
+		}
+	}
+
+	if m.canPoll(ms.LastPolled) {
+		poll := m.state.NewTask("poll-messages", "Poll store for device management messages")
+		ts.WaitFor(poll)
+		ts.AddTask(poll)
+	}
+
+	if len(ts.Tasks()) > 0 {
+		chg := m.state.NewChange(deviceMgmtCycleChangeKind, "Process device management messages")
+		chg.AddAll(ts)
+		chg.Set("dispatch-queued", dispatch != nil)
+	}
+
+	return nil
+}
+
+// canPoll checks whether polling should happen now based on feature flags and timing.
+func (m *DeviceMgmtManager) canPoll(lastPolled time.Time) bool {
+	tr := config.NewTransaction(m.state)
+	enabled, err := features.Flag(tr, features.MessagePolling)
+	if err != nil && !config.IsNoOption(err) {
+		logger.Noticef("cannot check message-polling feature flag: %v", err)
+		return false
+	}
+	if !enabled {
+		return false
+	}
+
+	if time.Since(lastPolled) < pollingInterval {
+		return false
+	}
+
+	return true
+}
+
+// doPollMessages polls the store for new request messages, acknowledges receipt
+// of persisted messages, and sends queued response messages.
+func (m *DeviceMgmtManager) doPollMessages(t *state.Task, tomb *tomb.Tomb) error {
+	m.state.Lock()
+
+	ms, err := m.getState()
+	if err != nil {
+		return err
+	}
+
+	deviceCtx, err := snapstate.DevicePastSeeding(m.state, nil)
+	if err != nil {
+		m.state.Unlock()
+		return err
+	}
+
+	sto := snapstate.Store(m.state, deviceCtx)
+	m.state.Unlock()
+
+	messages := make([]store.Message, 0, len(ms.ReadyResponses))
+	for _, msg := range ms.ReadyResponses {
+		messages = append(messages, msg)
+	}
+
+	resp, err := sto.PollMessages(
+		tomb.Context(context.Background()),
```

Comment:
It might be safer to do the usual `defer st.Unlock` at the top and just have a pair of `Unlock/Lock` around this call